### ISSUE TITLE
How to use Phi3 ONNX model in Triton efficiently?

### ISSUE THREAD

**Issue by khaerensml6** [microsoft/onnxruntime-genai#674]
I'm using the phi3 ONNX model in a triton model server.
It's running significantly slower than the pytorch model, probably because I'm making some obvious mistakes in setting this up. Any help would be welcome.

I want to set up the triton server to take in a user prompt as text and return the generated text, to this end I'm using a python backend that does the following:

tokenize the input text, create input ids tensor using hf transformers
create the attention mask tensor
create zero-initialized past_key_values tensors
then, repeat until EOS token is detected:

send over input_ids, attention_mask and past_key_values to the Phi3 onnx model
take the argmax of the returned logits and add to input ids
copy the returned present_key_values into new past_key_values tensors
In a diagram:

![image](https://github.com/microsoft/onnxruntime-genai/assets/92426912/ad369d9e-c8c1-46fe-a82a-ad13a69f8870)

Some questions I have:

I'm currently using the Phi3 model in Triton by simply putting the model.onnx and respective model.onnx.data in the Triton model repository. I can't seem to figure out how to use all the other configs and python files included in the ONNX repository in Triton; am I missing something here?

I'm "manually" copying the present_key_values into new past_key_values tensors. This seems very inefficient. I'm not even sure the Phi3 model is using the past_key_values for anything, since on the first call I'm sending over zeros, and that works fine, which seems strange to me.

Any pointers would be helpful, this is my first time setting this up, so I'm probably doing things wrong.

**Comment by natke** [comment]
Hi @khaerensml6, you are likely not doing anything wrong. The API is not directly integrated into the Triton backend yet. The inefficiencies that you are seeing is likely as a result of not being able to take advantage of the KV cache.

Do you have a production scenario that you are targeting?

**Comment by prashantaithal** [comment]
Does the onnx model contain kv cache or should the developer implement the kv cache in ORT outside of the model?

**Comment by baijumeswani** [comment]
> Does the onnx model contain kv cache or should the developer implement the kv cache in ORT outside of the model?

The KV cache management is done inside onnxruntime-genai. From the description you have shared, it seems you're not using onnxruntime-genai, but just the hosted phi3 onnx model?

Could you please clarify how (if at all) you're using the onnxruntime-genai in the triton server?

**Comment by baijumeswani** [comment]
[khaerensml6](https://github.com/khaerensml6) From the description you have shared, it seems you're not using onnxruntime-genai, but just the hosted phi3 onnx model?

Could you please clarify how (if at all) you're using the onnxruntime-genai in the triton server?

**Comment by khaerensml6** [comment]
@baijumeswani I want to host multiple models in a triton server, so I directly used the ONNX weights of the Phi3 model. This seems to be rather inneficient though. Afterwards I tried to just run the onnxruntime-genai library _inside_ the Triton server, but I feel this also isn't performing that well. Currenlty trying out VLLM

**Comment by jason-engage** [comment]
@khaerensml6 did you ever come up with a good solution for running phi 3 inside of Triton? I'm using it for another model and I'd like it to manage all my inference. I'm trying to get an ensemble working for phi 3.5 vision and hitting some hurdles, maybe related to a lack of KV cache and some time of strict onnxruntime-genai requirement.
